File: src/com/designing/singleton/java/Singleton_DC.java
package com.designing.singleton.java;

/**
 *
 * 单例模式
 * ->
 * 某个类 只能有一个实例 提供一个全局访问点
 *
 * 双重校验锁
 *  线程安全 能延时加载 以时间换空间
 */
public class Singleton_DC {
    private volatile static Singleton_DC instance = null;
    // 私有构造方法  防止外部实例化
    private Singleton_DC(){}
    public static Singleton_DC getInstance(){
        if (instance == null){
            synchronized (Singleton_DC.class){
                if (instance == null){
                    instance = new Singleton_DC();
                }
            }
        }
        return instance;
    }
}

File: src/com/designing/singleton/kotlin/Singleton_DC.kt
package com.designing.singleton.kotlin

// 双重校验锁
class Singleton_DC private constructor(){
    companion object {
        val instance : Singleton_DC by lazy (mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            Singleton_DC()
        }
    }
}
// 双重校验锁
class Singleton_DC2 private constructor(private val property : Int){
    companion object {
        @Volatile private var instance : Singleton_DC2? = null
        fun getInstance(property: Int) =
                instance ?: synchronized(this){
                    instance ?: Singleton_DC2(property).also { instance = it }
                }
    }
}